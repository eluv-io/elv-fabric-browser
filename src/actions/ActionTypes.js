const ActionTypes = {
  accessGroups: {
    list: "LIST_ACCESS_GROUPS",
  },
  accounts: {
    get: {
      currentAccountAddress: "GET_CURRENT_ACCOUNT_ADDRESS"
    }
  },
  content: {
    libraries: {
      list: "LIST_LIBRARIES",
      get: "GET_LIBRARY",
      groups: "GET_LIBRARY_GROUPS",
      types: "GET_LIBRARY_TYPES"
    },
    objects: {
      list: "LIST_OBJECTS",
      get: "GET_OBJECTS",
      versions: "GET_OBJECT_VERSIONS"
    },
    types: {
      get: "GET_TYPE",
      list: "LIST_TYPES"
    }
  },
  contracts: {
    list: "LIST_CONTRACTS",
    compile: "COMPILE_CONTRACTS",
    deployed: {
      list: "LIST_DEPLOYED_CONTRACTS",
      balance: "GET_CONTRACT_BALANCE",
      events: "GET_CONTRACT_EVENTS",
      clearEvents: "CLEAR_CONTRACT_EVENTS",
      call: "CALL_CONTRACT_METHOD",
    },
    error: {
      compile: "ERROR_CONTRACT_COMPILATION"
    }
  },
  logs: {
    clear: "CLEAR_BLOCKCHAIN_LOGS",
    list: "LIST_BLOCKCHAIN_LOGS"
  },
  notifications: {
    error: "SET_ERROR_MESSAGE",
    notification: "SET_NOTIFICATION_MESSAGE",
    clear: "CLEAR_NOTIFICATIONS"
  },
  requests: {
    status: {
      submitted: "REQUEST_SUBMITTED",
      completed: "REQUEST_COMPLETED",
      error: "REQUEST_ERROR"
    }
  },
  routes: {
    path: "ROUTING_FRAME_PATH",
    synchronize: "START_ROUTE_SYNCHRONIZATION"
  }
};

export default ActionTypes;
