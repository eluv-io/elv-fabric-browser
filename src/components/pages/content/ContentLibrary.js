import React from "react";
import {LibraryCard} from "../../components/DisplayCards";
import { Link } from "react-router-dom";
import Path from "path";
import ContentIcon from "../../../static/icons/content.svg";
import RequestPage from "../RequestPage";
import { LabelledField } from "../../components/LabelledField";
import Redirect from "react-router/es/Redirect";
import Fabric from "../../../clients/Fabric";
import ClippedText from "../../components/ClippedText";

class ContentLibrary extends React.Component {
  constructor(props) {
    super(props);

    const libraryId = this.props.libraryId || this.props.match.params.libraryId;

    this.state = {
      libraryId,
      visibleItems: {},
      isContentSpaceLibrary: Fabric.utils.EqualHash(Fabric.contentSpaceId, libraryId)
    };

    this.PageContent = this.PageContent.bind(this);
    this.RequestComplete = this.RequestComplete.bind(this);
  }

  componentDidMount() {
    this.setState({
      requestId: this.props.WrapRequest({
        todo: async () => {
          await this.props.SetCurrentAccount();
          await this.props.GetContentLibrary({libraryId: this.state.libraryId});
          await this.props.ListContentObjects({libraryId: this.state.libraryId});

          if(!this.state.isContentSpaceLibrary) {
            await this.props.ListContentLibraryGroups({libraryId: this.state.libraryId});
          }
        }
      })
    });
  }

  RequestComplete() {
    if(this.state.deleting && this.props.requests[this.state.requestId].completed) {
      this.setState({
        deleted: true
      });
    }

    this.setState({
      library: this.props.libraries[this.state.libraryId]
    });
  }

  DeleteContentLibrary() {
    if(confirm("Are you sure you want to delete this library?")) {
      this.setState({
        requestId: this.props.WrapRequest({
          todo: async () => {
            await this.props.DeleteContentLibrary({libraryId: this.state.libraryId});
          }
        }),
        deleting: true
      });
    }
  }

  DeleteContentLibraryButton() {
    // Don't allow deletion of special content space library
    if(this.state.isContentSpaceLibrary) { return null; }

    return (
      <button className="action delete-action" onClick={() => this.DeleteContentLibrary()}>
        Delete Content Library
      </button>
    );
  }

  ContentObjects() {
    let objectElements = [];

    for(const objectId of this.state.library.objects.sort()) {
      const object = this.props.objects[objectId];
      const infoText = this.state.isContentSpaceLibrary ? "" : object.status.description;

      objectElements.push(
        <LibraryCard
          key={object.id}
          libraryId={object.objectId}
          link={Path.join(this.props.match.url, object.id)}
          icon={object.imageUrl || ContentIcon}
          name={object.name}
          isOwner={object.isOwner}
          infoText={infoText}
          description={object.description}
          title={object.name}
        />
      );
    }

    return (
      <div className="object-info label-box">
        <h3>Content</h3>
        {objectElements}
      </div>
    );
  }

  ToggleButton(label, id) {
    const visible = this.state.visibleItems[id];
    const toggleVisible = () => this.setState({
      visibleItems: {
        ...this.state.visibleItems,
        [id]: !visible
      }
    });
    const toggleButtonText = (visible ? "Hide " : "Show ") + label;

    return (
      <div className="actions-container">
        <button
          className={"action action-compact action-wide " + (visible ? "" : "secondary")}
          onClick={toggleVisible}>
          { toggleButtonText }
        </button>
      </div>
    );
  }

  LibraryGroups() {
    if(this.state.isContentSpaceLibrary) { return null; }

    // Return if no groups exist
    if(!Object.values(this.state.library.groups).some(groups => groups.length > 0)) { return null; }

    return (
      <div>
        <h3>Groups</h3>
        {
          Object.keys(this.state.library.groups).map(groupType => {
            const groups = this.state.library.groups[groupType];

            // Display nothing if no groups of this type exist
            if (groups.length === 0) {
              return null;
            }

            return (
              <div className="library-groups-info indented" key={"library-group-" + groupType}>
                <h4>{groupType.capitalize()}</h4>
                {
                  groups.map(group => {
                    return (
                      <div className="group-info indented" key={"library-group-" + groupType + "-" + group.address}>
                        <h4>{group.name || "Unknown Group"}</h4>
                        <LabelledField label="Address" value={group.address}/>
                      </div>
                    );
                  })
                }

              </div>
            );
          })
        }
      </div>
    );
  }

  LibraryMetadata() {
    let metadata;
    if(this.state.visibleItems["meta"]) {
      metadata = <pre className="content-object-data">{JSON.stringify(this.state.library.meta, null, 2)}</pre>;
    }

    let privateMetadata;
    if(this.state.visibleItems["privateMeta"]) {
      privateMetadata = <pre className="content-object-data">{JSON.stringify(this.state.library.privateMeta, null, 2)}</pre>;
    }

    return [
      <div className="formatted-data" key="meta">
        <LabelledField label="Public Metadata" value={this.ToggleButton("Metadata", "meta")} />
        { metadata }
      </div>,
      <div className="formatted-data" key="privateMeta">
        <LabelledField label="Private Metadata" value={this.ToggleButton("Metadata", "privateMeta")} />
        { privateMetadata }
      </div>
    ];
  }

  LibraryImage() {
    if(this.state.library.imageUrl) {
      return (
        <div className="object-image">
          <img src={this.state.library.imageUrl} />
        </div>
      );
    }

    return null;
  }

  LibraryInfo() {
    const description = <ClippedText className="object-description" text={this.state.library.description} />;
    const libraryObjectPath = Path.join(this.props.match.url, this.state.library.libraryObjectId);

    return (
      <div className="object-info label-box">
        { this.LibraryImage() }
        <h3>Content Library Info</h3>
        <LabelledField label={"Library ID"} value={this.state.libraryId} />
        <LabelledField label={"Owner"} value={this.state.library.owner} />
        <LabelledField label={"Description"} value={description} />
        <LabelledField
          label={"Library Object"}
          value={
            <Link className="inline-link" to={libraryObjectPath}>
              {this.state.library.libraryObjectId}
            </Link>
          }
        />
        <LabelledField
          label={"Contract Address"}
          value={
            <Link className="inline-link" to={Path.join(libraryObjectPath, "contract")}>
              {this.state.library.contractAddress}
            </Link>
          }
        />
        <LabelledField label={"Content Objects"} value={this.state.library.objects.length} />
        { this.LibraryMetadata() }
        { this.LibraryGroups() }
      </div>
    );
  }

  Actions() {
    let manageGroupsButton;
    if(this.state.library.isOwner && !this.state.isContentSpaceLibrary) {
      manageGroupsButton = <Link to={Path.join(this.props.match.url, "groups")} className="action" >Manage Groups</Link>;
    }

    return (
      <div className="actions-container">
        <Link to={Path.dirname(this.props.match.url)} className="action secondary" >Back</Link>
        <Link to={Path.join(this.props.match.url, "edit")} className="action" >Edit Content Library</Link>
        { manageGroupsButton }
        <Link to={Path.join(this.props.match.url, "create")} className="action" >
          { this.state.isContentSpaceLibrary ? "New Content Type" : "New Content Object" }
        </Link>
        { this.DeleteContentLibraryButton() }
      </div>
    );
  }

  PageContent() {
    if(this.state.deleted) {
      return <Redirect push to={"/content"}/>;
    }

    return (
      <div className="page-container contents-page-container">
        { this.Actions() }
        <div className="object-display">
          <h3 className="page-header">{ this.state.library.name }</h3>
          { this.LibraryInfo() }
          { this.ContentObjects() }
        </div>
      </div>
    );
  }

  render() {
    return (
      <RequestPage
        pageContent={this.PageContent}
        requestId={this.state.requestId}
        requests={this.props.requests}
        OnRequestComplete={this.RequestComplete}
      />
    );
  }
}

export default ContentLibrary;
