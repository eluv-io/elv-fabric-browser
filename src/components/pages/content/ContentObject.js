import React from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import Path from "path";
import PrettyBytes from "pretty-bytes";
import { LabelledField } from "../../components/LabelledField";
import DashVideo from "../DashVideo";
import Redirect from "react-router/es/Redirect";
import ClippedText from "../../components/ClippedText";
import {PageHeader} from "../../components/Page";
import {DownloadFromUrl} from "../../../utils/Files";
import FileBrowser from "../../components/FileBrowser";
import AppFrame from "../../components/AppFrame";
import Fabric from "../../../clients/Fabric";
import Tabs from "elv-components-js/src/components/Tabs";
import Action from "elv-components-js/src/components/Action";
import {AccessChargeDisplay} from "../../../utils/Helpers";
import LoadingElement from "elv-components-js/src/components/LoadingElement";

class ContentObject extends React.Component {
  constructor(props) {
    super(props);

    const typeInfo = props.object.typeInfo || {};

    let initialView = "info";
    const displayAppUrl = props.object.displayAppUrl || (typeInfo.displayAppUrl);

    if(props.view) {
      initialView = props.view;
    } else if((displayAppUrl || props.object.videoUrl) && !props.object.isContentType) {
      //initialView = "display";
    }

    this.state = {
      visibleVersions: {},
      appRef: React.createRef(),
      view: initialView,
      displayAppUrl,
      typeHash: typeInfo.hash,
      typeName: (typeInfo.meta && typeInfo.meta.name) ? typeInfo.meta.name : typeInfo.hash
    };

    this.PageContent = this.PageContent.bind(this);
    this.SubmitContentObject = this.SubmitContentObject.bind(this);
  }

  async SubmitContentObject(confirmationMessage) {
    if(confirm(confirmationMessage)) {
      await this.props.methods.PublishContentObject({objectId: this.props.objectId});

      await this.props.Load({componentParams: {view: "info"}});
    }
  }

  DeleteContentObject() {
    if(confirm("Are you sure you want to delete this content object?")) {
      this.props.methods.DeleteContentObject({
        libraryId: this.props.libraryId,
        objectId: this.props.objectId
      });
    }
  }

  DeleteVersionButton(versionHash) {
    // Don't allow deleting of last version
    if(this.props.object.versions.length === 1) { return null; }

    const DeleteVersion = async () => {
      if(confirm("Are you sure you want to delete this content version?")) {
        await this.props.methods.DeleteContentVersion({
          libraryId: this.props.libraryId,
          objectId: this.props.objectId,
          versionHash
        });

        await this.props.Load({componentParams: {view: "info"}});
      }
    };

    return (
      <LoadingElement loading={this.props.methodStatus.DeleteContentVersion.loading} loadingIcon="rotate" >
        <Action
          className="delete-action action-compact action-wide"
          onClick={DeleteVersion}>
          Delete Content Version
        </Action>
      </LoadingElement>
    );
  }

  ToggleVersion(hash) {
    this.setState({
      visibleVersions: {
        ...this.state.visibleVersions,
        [hash]: !this.state.visibleVersions[hash]
      }
    });
  }

  ToggleButton(label, id) {
    const toggleVisible = () => this.ToggleVersion(id);
    const visible = this.state.visibleVersions[id];
    const toggleButtonText = (visible ? "Hide " : "Show ") + label;

    return (
      <Action className={"action-compact action-wide " + (visible ? "" : "secondary")} onClick={toggleVisible}>
        { toggleButtonText }
      </Action>
    );
  }

  PublishButton() {
    if(!this.props.object.isNormalObject) { return null; }

    const reviewerGroups = this.props.library.groups.reviewer;
    const reviewRequired = reviewerGroups.length > 0;

    if(this.props.object.status.code < 0 && this.props.object.isOwner) {
      const actionText = reviewRequired ? "Submit for Review" : "Publish";
      const confirmationMessage = reviewRequired ?
        "Are you sure you want to submit this content object for review?" :
        "Are you sure you want to publish this content object?";

      return (
        <Action onClick={() => this.SubmitContentObject(confirmationMessage)} >
          { actionText }
        </Action>
      );
    }

    if(this.props.object.status.code > 0 && this.props.object.permissions.canReview) {
      return (
        <Action type="link" to={Path.join(this.props.match.url, "review")}>
          Review
        </Action>
      );
    }
  }

  VersionSize(version) {
    return PrettyBytes(version.parts.reduce((a, part) => a + part.size, 0));
  }

  ToggleSection(label, id, value, format=true) {
    const visible = this.state.visibleVersions[id];

    let content;
    if(visible) {
      if(format) {
        content = <pre className="content-object-data">{JSON.stringify(value, null, 2)}</pre>;
      } else { content = value; }
    }

    return (
      <div className="formatted-data">
        <LabelledField label={label} value={this.ToggleButton(label, id)} />
        { content }
      </div>
    );
  }

  ObjectMedia() {
    let image;
    let video;

    if(this.props.object.imageUrl) {
      image = (
        <div className="object-image">
          <img src={this.props.object.imageUrl} />
        </div>
      );
    }

    if(this.props.object.meta["offering.en"]) {
      video = (
        <div className="object-video">
          <DashVideo videoUrl={this.props.object.imageUrl} />
        </div>
      );
    }

    if(!image && !video) {
      return null;
    } else {
      return (
        <div className="object-media">
          {video}
          {image}
        </div>
      );
    }
  }

  ObjectParts(version) {
    if(!version.parts || version.parts.length === 0) { return null; }

    let partNames = this.props.object.meta["eluv-fb.parts"] || {};
    let names = {};
    Object.keys(partNames).forEach(name => {
      names[partNames[name]] = name;
    });

    const parts = (version.parts.map((part, partNumber) => {
      const downloadButton = (
        <Action
          className="action-compact tertiary"
          onClick={() => {
            this.props.DownloadPart({
              libraryId: this.props.libraryId,
              objectId: this.props.objectId,
              versionHash: version.hash,
              partHash: part.hash,
              callback: async (url) => {
                await DownloadFromUrl(url, part.hash);
              }
            });
          }}
        >
          Download
        </Action>
      );

      const name = names[part.hash] ? <LabelledField label="Name" value={names[part.hash]}/> : null;

      return (
        <div key={part.hash + "-" + partNumber} className="part-info">
          { name }
          <LabelledField label="Hash" value={part.hash} />
          <LabelledField label="Size" value={PrettyBytes(part.size)} />
          <LabelledField label="" value={downloadButton} />
        </div>
      );
    }));

    return (
      <div>
        <h3>Parts</h3>
        { parts }
      </div>
    );
  }

  ObjectFiles() {
    const downloadMethod = async (filePath) => {
      await this.props.DownloadFile({
        libraryId: this.props.libraryId,
        objectId: this.props.objectId,
        filePath
      });
    };

    const uploadMethod = async (path, fileList) => {
      await this.props.methods.UploadFiles({
        libraryId: this.props.libraryId,
        objectId: this.props.objectId,
        path,
        fileList
      });
    };

    const urlMethod = async (filePath) => {
      return await this.props.FileUrl({
        libraryId: this.props.libraryId,
        objectId: this.props.objectId,
        filePath
      });
    };

    return(
      <div className="object-files">
        <h3>Files</h3>
        <FileBrowser
          files={this.props.object.meta.files || {}}
          Reload={() => this.props.Load({componentParams: {view: "files"}})}
          uploadStatus={this.props.methodStatus.UploadFiles}
          Upload={uploadMethod}
          Download={downloadMethod}
          FileUrl={urlMethod}
        />
      </div>
    );
  }

  ObjectVersion(version, versionNumber, latestVersion=false) {
    const visible = latestVersion || this.state.visibleVersions[version.hash];

    let versionHeader;
    if(!latestVersion) {
      versionHeader = <LabelledField label={"Version " + versionNumber} value={this.ToggleButton("Version", version.hash)} />;
    } else {
      versionHeader = <h3>{"Latest Version"}</h3>;
    }

    if(!visible) {
      return (
        <div key={"version-" + versionNumber} className="version-info indented">
          { versionHeader }
        </div>
      );
    }

    return (
      <div key={"version-" + versionNumber} className={"version-info " + (latestVersion ? "" : "indented version-visible")}>
        { versionHeader }

        <div className="indented">
          <LabelledField label={"Hash"} value={version.hash} />
          <LabelledField label={"Parts"} value={version.parts.length} />
          <LabelledField label={"Total size"} value={this.VersionSize(version)} />
          { this.ToggleSection("Metadata", "metadata-" + versionNumber, version.meta) }
          { this.ToggleSection("Verification", "verification-" + versionNumber, version.verification) }
          { this.ToggleSection("Parts", "parts-" + versionNumber, this.ObjectParts(version), false) }

          <br/>

          <LabelledField label="" value={this.DeleteVersionButton(version.hash)}/>
        </div>
      </div>
    );
  }

  PreviousVersions() {
    if(this.props.object.versions.length < 2) { return null; }

    return (
      <div>
        <h3>Previous Versions</h3>

        { this.props.object.versions.slice(1).map((version, i) => {
          const versionNumber = (i+1 - this.props.object.versions.length) * -1;
          return this.ObjectVersion(version, versionNumber);
        })}
      </div>
    );
  }

  ContractInfo() {
    let contractInfo = [];
    contractInfo.push(
      <LabelledField
        key={"contract-" + this.props.object.contractAddress}
        label={"Contract Address"}
        value={
          <Link className="inline-link" to={Path.join(this.props.match.url, "contract")}>
            {this.props.object.contractAddress}
          </Link>
        } />

    );

    if(!this.props.object.isContentLibraryObject && this.props.object.customContractAddress) {
      const customContractAddress = this.props.object.customContractAddress;

      contractInfo.push(
        <LabelledField
          key={"contract-" + customContractAddress}
          label={"Custom Contract"}
          value={<Link className="inline-link" to={Path.join(this.props.match.url, "custom-contract")}>{customContractAddress}</Link>} />);
    }

    return contractInfo;
  }

  // Display object status and review/submit/publish actions
  ObjectStatus() {
    if(!this.props.object.isNormalObject) { return null; }

    let reviewNote, reviewer;
    if(this.props.object.meta["eluv.reviewNote"]) {
      reviewer = <LabelledField label="Reviewer" value={this.props.object.meta["eluv.reviewer"]} />;
      const note = <ClippedText className="object-description" text={this.props.object.meta["eluv.reviewNote"]} />;
      reviewNote = <LabelledField
        label="Review Note" value={note} />;
    }

    return (
      <div>
        <LabelledField label={"Status"} value={this.props.object.status.description} />
        { reviewer }
        { reviewNote }
      </div>
    );
  }

  ObjectInfo() {
    const latestVersion = this.props.object.versions[0];
    const description = <ClippedText className="object-description" text={this.props.object.description} />;
    let accessCharge;
    if(this.props.object.isNormalObject) {
      accessCharge = <LabelledField label={"Access charge"} value={AccessChargeDisplay(this.props.object.baseAccessCharge)} />;
    }

    return (
      <div className="object-info label-box">
        <LabelledField label={"Name"} value={this.props.object.name} />
        <LabelledField label={"Description"} value={description} alignTop={true} />
        <br />
        { this.ObjectStatus() }
        { accessCharge }
        <br />
        <LabelledField label={"Library ID"} value={this.props.libraryId} />
        <LabelledField label={"Object ID"} value={this.props.objectId} />
        <LabelledField label={"Type"} value={this.state.typeName} hidden={this.props.object.isContentType} />
        <LabelledField label={"Type Hash"} value={this.state.typeHash} hidden={this.props.object.isContentType} />
        { this.ContractInfo() }
        <LabelledField label={"Owner"} value={this.props.object.owner} />
        <br />
        <LabelledField label={"Versions"} value={this.props.object.versions.length} />
        <LabelledField label={"Parts"} value={latestVersion.parts.length} />
        <LabelledField label={"Total size"} value={this.VersionSize(latestVersion)} />

        { this.ObjectVersion(this.props.object.versions[0], this.props.object.versions.length, true) }

        { this.PreviousVersions() }
      </div>
    );
  }

  Actions() {
    let manageAppsLink;
    if(this.props.object.isOwner) {
      manageAppsLink = (
        <Action type="link" to={Path.join(this.props.match.url, "apps")}>
          Apps
        </Action>
      );
    }

    let setContractButton;
    if(
      !this.props.object.customContractAddress &&
      (this.props.object.isNormalObject || this.props.object.isContentType) &&
      this.props.object.isOwner
    ) {
      setContractButton = (
        <Action type="link" to={Path.join(this.props.match.url, "deploy")}>
          Custom Contract
        </Action>
      );
    }

    let deleteObjectButton;
    if(!this.props.object.isContentLibraryObject) {
      deleteObjectButton = (
        <Action className="delete-action" onClick={() => this.DeleteContentObject()}>
          Delete
        </Action>
      );
    }

    return (
      <div className="actions-container">
        <Action type="link" to={Path.dirname(this.props.match.url)} className="secondary">Back</Action>
        <Action type="link" to={Path.join(this.props.match.url, "edit")}>Manage</Action>
        { this.PublishButton() }
        { setContractButton }
        <Action type="link" to={Path.join(this.props.match.url, "upload")}>Upload Parts</Action>
        { manageAppsLink }
        { deleteObjectButton }
      </div>
    );
  }

  DisplayVideo() {
    return (
      <div className="video-player">
        <video poster={this.props.object.imageUrl} controls={true}>
          <source src={this.props.object.videoUrl} />
        </video>
      </div>
    );
  }

  AppFrame() {
    if(!this.state.displayAppUrl) { return null; }

    const queryParams = {
      contentSpaceId: Fabric.contentSpaceId,
      libraryId: this.props.libraryId,
      objectId: this.props.objectId,
      type: this.state.typeHash,
      action: "display"
    };

    return (
      <AppFrame
        className="display-frame"
        appUrl={this.state.displayAppUrl}
        queryParams={queryParams}
        onComplete={this.LoadObject}
        onCancel={() => this.setState({deleted: true})}
      />
    );
  }

  Tabs() {
    let tabOptions = [
      ["Content Info", "info"],
      ["Files", "files"]
    ];

    if(this.state.displayAppUrl || this.props.object.videoUrl) {
      tabOptions.unshift(["Display", "display"]);
    }

    return (
      <Tabs
        options={tabOptions}
        selected={this.state.view}
        onChange={(value) => this.setState({view: value})}
      />
    );
  }

  PageContent() {
    if(this.props.methodStatus.DeleteContentObject.completed) {
      return <Redirect push to={Path.dirname(this.props.match.url)} />;
    }

    const header = this.props.object.isContentLibraryObject ?
      this.props.library.name + " > Library Object" :
      this.props.library.name + " > " + this.props.object.name;

    let pageContent;
    if(this.state.view === "display") {
      if(this.state.displayAppUrl) {
        pageContent = this.AppFrame();
      } else {
        pageContent = this.DisplayVideo();
      }
    } else if(this.state.view === "info") {
      pageContent = (
        <div>
          { this.ObjectMedia() }
          { this.ObjectInfo() }
        </div>
      );
    } else {
      pageContent = this.ObjectFiles();
    }

    return (
      <div className="page-container content-page-container">
        { this.Actions() }
        <PageHeader header={header} subHeader={this.props.object.description} />
        { this.Tabs() }
        <div className="page-content">
          { pageContent }
        </div>
      </div>
    );
  }

  render() {
    const loading = this.props.methodStatus.DeleteContentObject.loading || this.props.methodStatus.PublishContentObject.loading;

    return (
      <LoadingElement
        fullPage={true}
        loading={loading}
        render={this.PageContent}
      />
    );
  }
}

ContentObject.propTypes = {
  libraryId: PropTypes.string.isRequired,
  library: PropTypes.object.isRequired,
  objectId: PropTypes.string.isRequired,
  object: PropTypes.object.isRequired,
  methods: PropTypes.shape({
    UploadFiles: PropTypes.func.isRequired,
    DeleteContentObject: PropTypes.func.isRequired,
    DeleteContentVersion: PropTypes.func.isRequired
  }),
  DownloadPart: PropTypes.func.isRequired,
  DownloadFile: PropTypes.func.isRequired,
};

export default ContentObject;
