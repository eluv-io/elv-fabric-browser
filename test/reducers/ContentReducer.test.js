import ContentReducer from "../../src/reducers/ContentReducer";
import ActionTypes from "../../src/actions/ActionTypes";
import ContentLibrary from "../../src/models/ContentLibrary";
import ContentObject from "../../src/models/ContentObject";

describe("Request Reducer", () => {
  let initialState = {
    contentLibraries: {},
    contentObjects: {}
  };

  const contentLibrary = new ContentLibrary({
    libraryId: "ilibNLWSsptc11nHg6YNYhUair",
    contentObjectsData: [
      {
        "id": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
        "versions": [
          {
            "id": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
            "hash": "hq__QmWZ3FfjFio1TCuZjR5Qxpop6idvmzqW8LXJ1Ze2V648XS",
            "type": "",
            "meta": {
              "eluv.contract_address": "0xCe8f7fDc682C2b4Af0d412D5DAb046F820E25Ef9",
              "eluv.name": "TEST",
              "eluv.type": "library",
              "description": "test description"
            }
          }
        ]
      }
    ]
  });

  test("Initial state", () => {
    expect(ContentReducer(undefined, {type: "@@INIT"})).toEqual(initialState);
  });

  test("List all", () => {
    let resultState = ContentReducer(
      initialState,
      {
        type: ActionTypes.request.content.completed.list.all,
        contentLibraries: { "ilibNLWSsptc11nHg6YNYhUair": contentLibrary }
      }
    );

    expect(resultState).toMatchSnapshot();
  });

  test("Get library", () => {
    let resultState = ContentReducer(
      initialState,
      {
        type: ActionTypes.request.content.completed.list.library,
        libraryId: "ilibNLWSsptc11nHg6YNYhUair",
        contentLibrary: contentLibrary
      }
    );

    expect(resultState).toMatchSnapshot();
  });

  test("Get content object", () => {
    let contentObject = new ContentObject({
      libraryId: "ilibNLWSsptc11nHg6YNYhUair",
      contentObjectData: {
        "id": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
        "versions": [
          {
            "id": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
            "hash": "hq__QmWZ3FfjFio1TCuZjR5Qxpop6idvmzqW8LXJ1Ze2V648XS",
            "type": "",
            "meta": {
              "eluv.contract_address": "0xCe8f7fDc682C2b4Af0d412D5DAb046F820E25Ef9",
              "eluv.name": "TEST",
              "eluv.type": "library",
              "description": "test description"
            }
          }
        ]
      }
    });

    let resultState = ContentReducer(
      initialState,
      {
        type: ActionTypes.request.content.completed.list.contentObject,
        contentObject
      }
    );

    expect(resultState).toMatchSnapshot();
  });
});
