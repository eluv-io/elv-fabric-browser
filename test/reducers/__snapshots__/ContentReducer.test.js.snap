// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Reducer Get content object 1`] = `
Object {
  "contentLibraries": Object {},
  "contentObjects": Object {
    "hq__QmWZ3FfjFio1TCuZjR5Qxpop6idvmzqW8LXJ1Ze2V648XS": ContentObject {
      "description": "test description",
      "hash": "hq__QmWZ3FfjFio1TCuZjR5Qxpop6idvmzqW8LXJ1Ze2V648XS",
      "latestVersion": Object {
        "hash": "hq__QmWZ3FfjFio1TCuZjR5Qxpop6idvmzqW8LXJ1Ze2V648XS",
        "id": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
        "meta": Object {
          "description": "test description",
          "eluv.contract_address": "0xCe8f7fDc682C2b4Af0d412D5DAb046F820E25Ef9",
          "eluv.name": "TEST",
          "eluv.type": "library",
        },
        "type": "",
      },
      "libraryId": "ilibNLWSsptc11nHg6YNYhUair",
      "metadata": Object {
        "description": "test description",
        "eluv.contract_address": "0xCe8f7fDc682C2b4Af0d412D5DAb046F820E25Ef9",
        "eluv.name": "TEST",
        "eluv.type": "library",
      },
      "name": "TEST",
      "objectId": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
      "type": "library",
      "versions": Array [
        undefined,
      ],
    },
  },
}
`;

exports[`Request Reducer Get library 1`] = `
Object {
  "contentLibraries": Object {
    "ilibNLWSsptc11nHg6YNYhUair": ContentLibrary {
      "contentObjects": Array [],
      "description": "test description",
      "libraryId": "ilibNLWSsptc11nHg6YNYhUair",
      "libraryMetadataObject": ContentObject {
        "description": "test description",
        "hash": "hq__QmWZ3FfjFio1TCuZjR5Qxpop6idvmzqW8LXJ1Ze2V648XS",
        "latestVersion": Object {
          "hash": "hq__QmWZ3FfjFio1TCuZjR5Qxpop6idvmzqW8LXJ1Ze2V648XS",
          "id": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
          "meta": Object {
            "description": "test description",
            "eluv.contract_address": "0xCe8f7fDc682C2b4Af0d412D5DAb046F820E25Ef9",
            "eluv.name": "TEST",
            "eluv.type": "library",
          },
          "type": "",
        },
        "libraryId": "ilibNLWSsptc11nHg6YNYhUair",
        "metadata": Object {
          "description": "test description",
          "eluv.contract_address": "0xCe8f7fDc682C2b4Af0d412D5DAb046F820E25Ef9",
          "eluv.name": "TEST",
          "eluv.type": "library",
        },
        "name": "TEST",
        "objectId": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
        "type": "library",
        "versions": Array [
          undefined,
        ],
      },
      "name": "TEST",
    },
  },
  "contentObjects": Object {},
}
`;

exports[`Request Reducer List all 1`] = `
Object {
  "contentLibraries": Object {
    "ilibNLWSsptc11nHg6YNYhUair": ContentLibrary {
      "contentObjects": Array [],
      "description": "test description",
      "libraryId": "ilibNLWSsptc11nHg6YNYhUair",
      "libraryMetadataObject": ContentObject {
        "description": "test description",
        "hash": "hq__QmWZ3FfjFio1TCuZjR5Qxpop6idvmzqW8LXJ1Ze2V648XS",
        "latestVersion": Object {
          "hash": "hq__QmWZ3FfjFio1TCuZjR5Qxpop6idvmzqW8LXJ1Ze2V648XS",
          "id": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
          "meta": Object {
            "description": "test description",
            "eluv.contract_address": "0xCe8f7fDc682C2b4Af0d412D5DAb046F820E25Ef9",
            "eluv.name": "TEST",
            "eluv.type": "library",
          },
          "type": "",
        },
        "libraryId": "ilibNLWSsptc11nHg6YNYhUair",
        "metadata": Object {
          "description": "test description",
          "eluv.contract_address": "0xCe8f7fDc682C2b4Af0d412D5DAb046F820E25Ef9",
          "eluv.name": "TEST",
          "eluv.type": "library",
        },
        "name": "TEST",
        "objectId": "iq__S5Q8fUSmFbTyZ6D53HPCuR",
        "type": "library",
        "versions": Array [
          undefined,
        ],
      },
      "name": "TEST",
    },
  },
  "contentObjects": Object {},
}
`;
